### PR TITLE
Fix atmel-samd alarm module

### PR DESCRIPTION
**Pull request by t-ikegami**
This PR solves most of the TODO stuffs of atmel-samd alarm module.  Notable changes are:

- Make light sleep work with TimeAlarm (enable RTC_IRQ)
- PinAlarm sets up pins just before light sleep, and resets them after wakeup. [1]
- Support edge parameter of PinAlarm
- Support all tamper pins in PinAlarm for deep sleep.
- Implement SleepMemory
- Add board_deep_sleep_hook() function to be called just before deep sleep. [2]

[1] It is because PinAlarm does not have deinit.  
[2] It is an empty function by default.  In Wio Terminal, the hook function is used to turn off RTL8720 coprocessor.  (RTL8720 is enabled on board reset, and draws considerable amount of power.)

To my experiment, PinAlarm in deep sleep is triggered on the transition to Hi-Z state, not by edge.  Double-check may be necessary.

### CODE REVIEW

**Comment by tannewt** [comment]
> [2] It is an empty function by default. In Wio Terminal, the hook function is used to turn off RTL8720 coprocessor. (RTL8720 is enabled on board reset, and draws considerable amount of power.)

Did you try using `board_deinit` for this? I think it's meant to be used in the same way.

File: locale/circuitpython.pot [hunk 1]
```diff
@@ -2158,7 +2166,7 @@ msgstr ""
 msgid "Size not supported"
 msgstr ""
 
-#: ports/atmel-samd/common-hal/alarm/SleepMemory.c
+#: ports/atmel-samd/common-hal/alarm.orig/SleepMemory.c
```
Comment (tannewt) [inline locale/circuitpython.pot:2169]: I think changing the suffix of the file will make the translation scripts ignore it. (And therefore leave this as-is.)

File: locale/circuitpython.pot [hunk 2]
```diff
@@ -1931,6 +1935,10 @@ msgstr ""
 msgid "Pin is input only"
 msgstr ""
 
+#: ports/atmel-samd/common-hal/alarm/pin/PinAlarm.c
+msgid "Pin is not free"
```
Comment (tannewt) [inline locale/circuitpython.pot:1939]: Is there a `Pin in use` message you can use instead? Adding another message takes more space and requires new translations.

File: ports/atmel-samd/common-hal/alarm/__init__.c [hunk 3]
```diff
@@ -103,64 +91,56 @@ mp_obj_t common_hal_alarm_create_wake_alarm(void) {
...
     mp_obj_t wake_alarm = mp_const_none;
 
+    // This works but achieves same power consumption as time.sleep()
+    PM->SLEEPCFG.reg = PM_SLEEPCFG_SLEEPMODE_STANDBY;
+    while (PM->SLEEPCFG.bit.SLEEPMODE != PM_SLEEPCFG_SLEEPMODE_STANDBY_Val) {
+    }
+    // Even though RAMCFG_OFF, SYSRAM seems to be retained.  Probably
+    // because RTC keeps sleepwalking.  Anyway, STDBYCFG should be
+    // left intact as 0 to retain SYSRAM.
+    #if 0
```
Comment (tannewt) [inline ports/atmel-samd/common-hal/alarm/__init__.c:107]: Please remove anything that is `#if 0`'d out. The git history has a copy if we need it.

File: ports/atmel-samd/common-hal/alarm/__init__.c [hunk 4]
```diff
@@ -245,17 +216,24 @@ void NORETURN common_hal_alarm_enter_deep_sleep(void) {
...
+    board_deep_sleep_hook();
+    port_idle_until_interrupt();
 }
+#endif
 
 void common_hal_alarm_gc_collect(void) {
     gc_collect_ptr(shared_alarm_get_wake_alarm());
 }
+
+MP_WEAK void board_deep_sleep_hook(void) {
```
Comment (tannewt) [inline ports/atmel-samd/common-hal/alarm/__init__.c:238]: Should we add this for all boards and declare it in `supervisor/board.h`?

**Review by tannewt — CHANGES_REQUESTED** [review]
Thank you for the PR! I've left a few comments.

@maholli you are using SAMD51 sleep as well right? Mind testing this PR?

**Comment by maholli** [comment]
@tannewt Sure I'll kick the tires on this PR tonight.

@t-ikegami your comment adding edge support for PinAlarm reminded me of this SAMD5x errata. Did you run into this?

**Comment by maholli** [comment]
Ok I played with the new SAMD alarm features implemented in this PR.

- [x] power consumption is on-par with what I was measuring with my initial samd Alarm implementation. Tested deep_sleep_until_alarms with `TimeAlarm`, `PinAlarm` and both at once. 
- [x] `alarm.wake_alarm` still properly distinguishes between time and pin alarms.
- [x] max deep sleep timealarm is still ~1500 days since prescalar is still 1024. That's good, thanks for not changing this. 
- [x] fake deep sleep still works as expected
- [ ] NOTE: I didn't try or test anything related to light sleep

Questions:

1. See [question about errata](https://github.com/adafruit/circuitpython/pull/6048#issuecomment-1043355688). I think this may cause trouble if a user attempts to wake on edge.
2. It looks for `SleepMemory` you're just using NVM starting at the BKUPRAM base address. It also looks like the size will default to the NVM's page size (8192 bytes) since that's what I got when I ran `len(alarm.sleep_memory[:])`. @tannewt can you confirm that CP doesn't need/use that address space? BKUPRAM starts at 0x47000000 per samd51j19a.h
3. It seems to me that your `board_deep_sleep_hook()` should really be done by the user in software (i.e. CP). If you want things turned off during deep sleep, that should be performed in software before calling `exit_and_deep_sleep_until_alarms` right??

File: locale/circuitpython.pot (hunk 1 above)
Comment (t-ikegami) [reply locale/circuitpython.pot:2169]: > @tannewt Sure I'll kick the tires on this PR tonight.
> 
> @t-ikegami your comment adding edge support for PinAlarm reminded me of this SAMD5x errata. Did you run into this? 

I have tried to check the errata condition, by using a Wio Terminal in hand and PA02 pin.  As far as wake up from BACKUP sleep is concerned, the tamper pin interrupt is triggered only for the transition from high/low to Hi-Z state, irrelevant to TAMLVL setting.  The behavior disagrees with the datasheet/errata, though it seems to be reasonable as a tamper detection.  Note that, while not in sleep mode (or IDLE sleep, maybe), the TAMPID register reacts as expected: if we set TAMLVLx = 0, TAMPID is flagged only on the falling edge.  Honestly, I'm confused, and hope someone to make a double check.

**Comment by t-ikegami** [comment]
> @tannewt Sure I'll kick the tires on this PR tonight.
> 
> @t-ikegami your comment adding edge support for PinAlarm reminded me of this SAMD5x errata. Did you run into this? 

I have tried to check the errata condition, by using a Wio Terminal in hand and PA02 pin.  As far as wake up from BACKUP sleep is concerned, the tamper pin interrupt is triggered only for the transition from high/low to Hi-Z state, irrelevant to TAMLVL setting.  The behavior disagrees with the datasheet/errata, though it seems to be reasonable as a tamper detection.  Note that, while not in sleep mode (or IDLE sleep, maybe), the TAMPID register reacts as expected: if we set TAMLVLx = 0, TAMPID is flagged only on the falling edge.  Honestly, I'm confused, and hope someone to make a double check.

**Comment by t-ikegami** [comment]
> 3. It seems to me that your `board_deep_sleep_hook()` should really be done by the user in software (i.e. CP). If you want things turned off during deep sleep, that should be performed in software before calling `exit_and_deep_sleep_until_alarms` right??

Unfortunately, cleanup_after_vm is called befre deep sleep, where pin states are reset.  Actually, I first tried the scheme, and got surprized that the power consumption bumps up in deep sleep.

**Comment by tannewt** [comment]
> 2\. @tannewt can you confirm that CP doesn't need/use that address space? BKUPRAM starts at 0x47000000 per samd51j19a.h

I don't believe we do. Sleep memory is the best use of this space anyway.

> Unfortunately, cleanup_after_vm is called befre deep sleep, where pin states are reset. Actually, I first tried the scheme, and got surprized that the power consumption bumps up in deep sleep.

My intention was that `board_deinit` would be used to put things in a low power state.

You can cheat a little bit for pin state by using PinAlarm to set pulls. We probably will want a proper pin state during sleep API in the future.

**Comment by tannewt** [comment]
Thanks for testing this @maholli!

**Comment by tannewt** [comment]
@t-ikegami Are you ready for another review? Let me know when you are.

**Comment by t-ikegami** [comment]
> @t-ikegami Are you ready for another review? Let me know when you are.

Please go ahead.

**Review by tannewt — APPROVED** [review]
Thank you for these updates! We can merge them into main for 7.3.0.

**Pull request merged**